text/0003-webcontents-invoke.md

```diff
@@ -0,0 +1,373 @@
+     > Current Solution: If there are no renderer process to respond, the promise will be rejected due to the timeout option in `webContents.invoke()`. This ensures that the user receives a response instead of hanging indefinitely. Default value of 30000 ms is used as default (value inspired by [libraries that interact with Chromium](https://pptr.dev/api/puppeteer.waittimeoutoptions/#timeout))
+
+2. Frame Navigation: If the target frame navigates before responding to the IPC message.
+
+   a. cross-site navigation: If a render frame navigates to a different site (not just a different page), it will be swapped out, and a new render frame host will be created (process isolation).
+
+   b. same-site navigation: If the frame navigates to a different page within the same site, the same render frame host will be used.
+
+
+      > Current Solution: Whether same or cross site navigation, the ipcRenderer will respond with a resolved value from the corresponding WebFrameMain/RenderFrameHost that made the call. Ideally we want to reject the response if it is not the same page we intended to invoke from the main process, but there is no way to control that same site navigation re-uses the same frame with the same context. 
+
+      >The target frame is decided when the main process makes the invoke because the WebFrameMain/RenderFrameHost corresponds to the ElectronRenderer/RendererAPI. So if the render frame navigates away after invoking the request, the promise will be resolved with the handler from the WebFrameMain/RenderFrameHost that made the call. 
+      
+     >The implementation checks that the routing ID of the main frame (RenderFrameHost) matches the routing ID of the web frame that received the invoke call. This ensures that the main process receives a response from the intended web frame and not a different one. However, this check may be redundant because the ElectronRenderer/RendererAPI always corresponds to the calling WebFrameMain/RenderFrameHost, so the routing IDs should never be different.
+
+3. No Registered Handler: If the renderer process does not have a registered handler for the specified channel, the promise should be rejected.
+
+   > Current Solution: When page changes, whether same or cross site navigation, the renderer process updates and new handlers are registered specific to that renderer process. So if a call from main is made before a handler registers or to a handler that does not exist, the promise will be rejected.
+
+### Edge Cases Not Addressed
+
+4. Multiple Frames: The `webContents.invoke()` method supports invoking from the mainFrame WebFrameMain. Future area of development is being able to invoke from a specific RenderFrameHost/ WebFrameMain instance by providing the routingId/ frameId option. This allows for communication with individual frames within a single webContents instance.
```

Comment:
question: would this be extensible via the existing shape of the API via the `options` parameter?

Reply:
Yes! If the user provides the `processId` and `routingId`, then we can get the webFrameMain with `webFrameMain.fromId(processId, routingId)` and send the message that way. Would be super extensible with the options for the user to provide the ids. There could be more edge cases with invoking a response from a non-main frame but theoritically, it would be easy to extend this API to send from any host frame. I'll add this to the rfc!